### ISSUE TITLE
Relax privilege requirements for org.freedesktop.resolve1.Manager.RegisterService

### ISSUE THREAD

**Issue by xytovl** [systemd/systemd#34248]
### Component

systemd-resolved

### Is your feature request related to a problem? Please describe

I develop an application that provides a service over the network and wants to advertise it: https://github.com/WiVRn/WiVRn

It is a streaming OpenXR runtime, has a client and server part which need to discover each other. In order to do so we use dns-sd and publish through avahi.

We would like to support systemd-resolved stack as well but #21669 is not implemented and we would have to use `org.freedesktop.resolve1.Manager`.
This API requires authorization (https://github.com/systemd/systemd/issues/29598#issuecomment-1958065776), I do not understand why this is considered a privileged operation.

### Describe the solution you'd like

Authorization for RegisterService/UnregisterService should not be enforced by default.

**Comment by bluca** [comment]
Authorization can be granted via polkit, have you tried that?

**Comment by xytovl** [comment]
Is it possible to add polkit rules for applications distributed through flatpak?

**Comment by bluca** [comment]
I don't know, that's something to ask in a flatpak support channel

**Comment by bluca** [comment]
Note that if you don't want this API to be gated on your system, you can simply have a polkit rule that always returns `polkit.Result.YES` for it, regardless of any other details

**Comment by xytovl** [comment]
https://lists.freedesktop.org/archives/flatpak/2018-August/001249.html

> However you are not allowed to install policy kit policy or rule
files, or system-bus dbus services, which is what creates the
privilege escalation step.

I think there is a gap in features:
My application provides a service over the network for virtual reality headsets to connect to. The reasonable technology for service discovery is mDNS/dns-sd.
With avahi-daemon, it is by default possible for a user program to publish a service and be discoverable.
Asking for (root) password is not acceptable for this use-case.

I understand the rationale of flatpak to not distribute polkit rules, as they would need to be manually audited.
I do not understand the rationale for privileges requirements on RegisterService API.

**Comment by evverx** [comment]
Port 5353 was chosen to make it possible for unprivileged processes to send/receive mDNS packets without having to be `root` (and they can do that by bypassing `resolved` altogether and doing that directly (though relatively recent changes in resolved managed to mess that up as well)).

> I do not understand the rationale for privileges requirements on RegisterService API.

I think it was implemented this way because use cases where a bunch of unprivileged processes advertise their stuff had never been considered (even though those use cases are the most popular use cases). Also it is much easier because the unprivileged counterpart would be much harder to implement.

> you can simply have a polkit rule that always returns polkit.Result.YES for it

This advice is really bad. The method isn't ready for any sort of unprivileged abuse. I don't think it even limits the number of services. (avahi wasn't ready to deal with things like https://github.com/avahi/avahi/issues/455 either but it was fixed).

**Comment by xytovl** [comment]
Is there a change I could propose as a pull request to allow unprivileged applications to publish services?

**Comment by niklaswimmer** [comment]
Hi, maybe there was misunderstanding, but the application does not need to be privileged to call this method. All that needs to be done is to set the correct message flag: `ALLOW_INTERACTIVE_AUTHORIZATION` (has a detailed description in the dbus spec). 

I tested the method with `busctl`, which opens up a dialog that asks me to enter my password (note that it does not ask for root access, just my user password) and says "Authentication is required to register a DNS-SD service".

```bash
busctl --system call \
      org.freedesktop.resolve1 \
      /org/freedesktop/resolve1 \
      org.freedesktop.resolve1.Manager \
      RegisterService \
      'sssqqqaa{say}' \
      "test-service" "test-service-name" "_http._tcp" 9999 0 0 0
```

<img width="2256" height="1504" alt="Image" src="https://github.com/user-attachments/assets/6450b9a3-4659-4084-be1f-45287a30313d" />

I have not tested it from a flatpak sandbox, but I see no reason why it would work differently there. 

From a user perspective, this is pretty much exactly what I _want_ to happen. I do not want my sandboxed application to just go and register DNS-SD services, without asking for my consent (DNS-SD has privacy implications for example, see [RFC 8882](https://www.rfc-editor.org/info/rfc8882/)).

If your application on the other hand is a system service, then it should not be a problem to include a polkit file (as described in a comment above) in whatever you use to distribute that service.

As such I think this issue can be closed.

**Comment by xytovl** [comment]
I don't understand why publishing mDNS services needs to be privileged. The application can use port 5353 and implement the mDNS responder directly.

Requesting administrator privilege each time the application start is not acceptable for users. My case is a VR streaming application, it needs to advertise over the network and runs as the user starting it.

With systemd-resolved and avahi being incompatible (they generate name conflicts if run together), this is going to make it more complicated for users

**Comment by RaitoBezarius** [comment]
It's up to distributions to decide their policy for that, no? I don't want to have services randomly register mDNS and sure, an app can use port 5353 and implement the mDNS responder, it doesn't mean my firewall will allow it.

While the situation may not be perfect right now, there's no reason to go in the other direction to signal that it's suddenly OK?

**Comment by xytovl** [comment]
Is there at least a mechanism to remember the authorization? I can't ask the user to enter the admin password each time they want to stream a game.

**Comment by niklaswimmer** [comment]
I agree with @RaitoBezarius that removing the authorisation is not the solution here.

I also agree however that asking the user for their password every time is not the best behaviour (even though, they are probably not asked for the administrator password, but for their password, which they use to unlock their device; at least that is how it works on my machine).

From my understanding, there is no mechanism to remember the authorisation. There seem to be some heuristics for when not to ask and instead reuse a previous authorisation but they definitely do not persist across sessions and probably not even across application restarts (I am not an expert on this, I can just tell you what I observed on my machine locally, where launching `busctl` from the same terminal multiple times would not trigger authorisation again, but running it from a new terminal window would).

I do not think this is an unsolvable problem, but from what I have read in the previous days (I am deep down the rabbit hole) the Linux desktop does not seem to be there yet. To me it seems that remembering the authorisation would at least require some kind of "application identity" to ensure the application is actually the same as the one the user authorised already (to me this sounds similar to what the https://github.com/linux-credentials/ group seems to need to map applications to origins, so we may have that at some point in the future). And then you'd of course also need all the glue to thread that information back into polkit.

You'd have to test under what conditions your users would have to re-authorise (is it every time you call the method, only when the application restarts, etc). If it is once per session, then it may be ok, or maybe not, that is up to you. I think as a sandboxed applicaton your only other options are to tell the user how they could modify their polkit files to allow your application (or you use the systemd d-bus to effectively call `run0` to create the polkit rule files directly from your app, see https://github.com/flatpak/flatpak/issues/6223; not sure if that is desirable though).

So, to reiterate, I do not believe that relaxing the privilege requirements is the solution, but I do also not believe that there is an actual good solution for your problem right now. Maybe just keep using Avahi for now 🤷

**Comment by xytovl** [comment]
The [screencast portal](https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.impl.portal.ScreenCast.html) handles a similar problem:
The request includes a persist_mode parameter, the response will tell if the requested mode was granted, and a token to authenticate later.
It's up to the application to store the token and provide it back later.
